kubernetes/client/kubernetes.go

```diff
@@ -0,0 +1,738 @@
+		if err != nil {
+			panic(err.Error())
+		}
+		// Set up internal clientset and clusterConfig
+		p.clientset = clientset
+		p.clusterConfig = clusterConfig
+		// Create REST client
+		p.RESTClient = clientset.CoreV1().RESTClient()
+		p.Logger.Info("succesfully authenticated using InClusterConfig()")
+		return clientset, clusterConfig, nil
+
+	}
+
+}
+
+// Initialize uses the passed clientset to create some authenticated clients
+// used in other methods. Creates a new namespace for wr to work in. Optionally
+// pass a namespace as a string.
+func (p *Kubernetesp) Initialize(clientset kubernetes.Interface, namespace ...string) error {
+
+	// If a namespace is passed, check it exists. If it does not, create it. If
+	// no namespace passed, create a random one.
+	if len(namespace) == 1 {
+		p.NewNamespaceName = namespace[0]
+		_, err := clientset.CoreV1().Namespaces().Get(p.NewNamespaceName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				err = p.CreateNewNamespace(p.NewNamespaceName)
+				if err != nil {
+					p.Logger.Crit(fmt.Sprintf("failed to create provided namespace: %s", err))
+				}
+			}
+		}
+	} else {
+		rand.Seed(time.Now().UnixNano())
+		p.NewNamespaceName = strings.Replace(namesgenerator.GetRandomName(0), "_", "-", -1) + "-wr"
+		p.Logger.Info(fmt.Sprintf("NewNamespaceName: %v \n", p.NewNamespaceName))
+		// Retry if namespace taken
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			nsErr := p.CreateNewNamespace(p.NewNamespaceName)
+			if nsErr != nil {
+				p.Logger.Warn("failed to create new namespace. Trying again.", "namespace", p.NewNamespaceName, "err", nsErr)
+				p.NewNamespaceName = strings.Replace(namesgenerator.GetRandomName(0), "_", "-", -1) + "-wr"
+			}
+			return nsErr
+		})
+		if retryErr != nil {
+			p.Logger.Error("creation of new namespace failed", "err", retryErr)
+			return fmt.Errorf("Creation of new namespace failed: %v", retryErr)
+		}
+	}
+
+	// Create client for deployments that is authenticated against the given
+	// cluster. Use default namsespace.
+	p.deploymentsClient = clientset.AppsV1beta1().Deployments(p.NewNamespaceName)
+
+	// Create client for services
+	p.serviceClient = clientset.CoreV1().Services(p.NewNamespaceName)
+
+	// Create client for pods
+	p.podClient = clientset.CoreV1().Pods(p.NewNamespaceName)
+
+	// Create configMap client
+	p.configMapClient = clientset.CoreV1().ConfigMaps(p.NewNamespaceName)
+
+	// Store md5 hashes of the data in each configMap. used when calling
+	// newConfigMap().
+	p.configMapHashes = map[string][16]byte{}
+
+	// ToDO: This assumes one portforward per namespaced deployment This should
+	// probably go in pod and the channels be created in an options struct to
+	// better isolate the logic
+
+	// Make channels for port forwarding
+	p.StopChannel = make(chan struct{}, 1)
+	p.ReadyChannel = make(chan struct{})
+
+	return nil
+}
+
+// Deploy creates the wr-manager deployment and service. Creates
```

Comment:
Err, not easily, with Deploy() doing everything it does. I imagine I might be able to write for loops in to deal with n env vars. It's wr specific because the `PodSpec` expects a binary from a tarball, that is obtained via an init container to be in a specific place. :(